### DISCUSSION TITLE
How to sign-in a new user programmatically, immediately after a user/pass sign-up?

### DISCUSSION THREAD

**Discussion by jayliew** [nextauthjs/next-auth#1574] [Help]
This question pertains to using the "Credentials" Provider (username/password), with JWT and database.

I'd like to create a signup/registration page, where immediately upon success, I automatically sign-in the new user. 

From what I gathered from the docs, to sign-in a user, I need to POST to the endpoint `/api/auth/signin/credentials` the values of:
- username
- password
- csrfToken
- callbackUrl

in a URL-encoded Form.

_Update: I have even tried POST'ing to the other end-point `/api/auth/callback/credentials` but still no luck._

Is this the right way to go about this? 

As a test, I tried accomplishing the above with a HTTP REST client (not a full-fledged browser) to see the response, and noticed that I receive a HTTP 302 that sends me back to the sign-in page at `/api/auth/signin`

I'm very new to Next Auth so I hope to learn the typical best practice of how to go about doing this.

The HTTP request:
```
> POST /api/auth/signin/credentials HTTP/1.1
> Host: localhost:3000
> Cookie: next-auth.session-token=eyJhbGciOiJkaXIiLCJlSNIPPEDhE5h0rSX8qX93SiY.EZo29bsXGcGsaNkLnVlr5g; next-auth.callback-url=http%3A%2F%2Flocalhost%3A3000; next-auth.csrf-token=7c172262SNIPPED6bd8f88e9fc7e749f58f6a9a6; __profilin=p%3Dt
> Content-Type: application/x-www-form-urlencoded
> Accept: */*
> Content-Length: 142

| username=uuuuuu&password=pppppppp&csrfToken=7c1722SNIPPED3d833c5132426e80fcfe5d3&callbackUrl=http%3A%2F%2Flocalhost%3A3000
```

The HTTP response:

```
< HTTP/1.1 302 Found
< Content-Type: text/html
< Location: http://localhost:3000/api/auth/signin
< Vary: Accept-Encoding
< Date: Thu, 25 Mar 2021 01:15:29 GMT
< Connection: keep-alive
< Transfer-Encoding: chunked

```
Which prompts the browser to then GET the sign-in page at `/api/auth/signin`

I thought that perhaps, the server response would send me to the callbackUrl I had specified in the POST?

**Comment by balazsorban44** [comment]
you will be redirected to the signin page, in case there was a problem with the CSRF token. make sure you send it correctly. we provide a `getCsrfToken` method for it. works both server and client side

https://next-auth.js.org/getting-started/client#getcsrftoken

**Reply by jayliew** [reply]
I have a bare-bones sign-in page that is pretty much as shown in the [docs](https://next-auth.js.org/configuration/pages#credentials-sign-in). It correctly gets the `csrfToken` (every time I refresh the page and look at the source, I can see the value changing), but when I submit the form, it doesn't log me in but instead send me to the default sign-in form.

I added a `console.log()` inside of `async authorize(credentials)` in `Providers.Credentials()`, and can confirm that my custom sign-in page isn't triggering `authorize()` at all.

![screenshot](https://i.imgur.com/XMu1kFn.png)


This is my bare-bones custom sign-in page:
```js
import { getCsrfToken } from 'next-auth/client'

export async function getStaticProps(context){
  const csrfToken = await getCsrfToken(context)
  return {
    props: { csrfToken }
  }
}

const Signup = ({ csrfToken }) => {
  return(
    <>
      <h1>Signup Page here</h1>
      <form action="/api/auth/callback/credentials" method="post">
        username: <input type="text" name="username" />
        <br />
        <br />
        password: <input type="password" name="password" />
        <br />
        <br />
        <input type="hidden" name="csrfToken" defaultValue={csrfToken} />
        <input type="hidden" name="callbackUrl" value="http://localhost:3000/" />
        <button type='submit'>Sign in</button>
      </form>
      <br />
    </>
  )
}

export default Signup;
```

The generated HTML source seems to contain the CSRF token and callback too:
```html
<!DOCTYPE html>
<html>
   <head>
      <style data-next-hide-fouc="true">body{display:none}</style>
      <noscript data-next-hide-fouc="true">
         <style>body{display:block}</style>
      </noscript>
      <meta charSet="utf-8"/>
      <meta name="viewport" content="width=device-width"/>
      <meta name="next-head-count" content="2"/>
      <noscript data-n-css=""></noscript>
      <link rel="preload" href="/_next/static/chunks/main.js?ts=1616720294380" as="script"/>
      <link rel="preload" href="/_next/static/chunks/webpack.js?ts=1616720294380" as="script"/>
      <link rel="preload" href="/_next/static/chunks/pages/_app.js?ts=1616720294380" as="script"/>
      <link rel="preload" href="/_next/static/chunks/pages/signup.js?ts=1616720294380" as="script"/>
      <noscript id="__next_css__DO_NOT_USE__"></noscript>
   </head>
   <body>
      <div id="__next">
         <h1>Signup Page here</h1>
         <form action="/api/auth/callback/credentials" method="post">username: <input type="text" name="username"/><br/><br/>password: <input type="password" name="password"/><br/><br/><input type="hidden" name="csrfToken" value="62495fce7cxxxxxxSNIPPEDxxxxxxxc1300cce788"/><input type="hidden" name="callbackUrl" value="http://localhost:3000/"/><button type="submit">Sign in</button></form>
         <br/>
      </div>
      <script src="/_next/static/chunks/react-refresh.js?ts=1616720294380"></script><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"csrfToken":"62495fce7cxxxxxxSNIPPEDxxxxxxxc1300cce788"},"__N_SSG":true},"page":"/signup","query":{},"buildId":"development","nextExport":false,"isFallback":false,"gsp":true}</script><script nomodule="" src="/_next/static/chunks/polyfills.js?ts=1616720294380"></script><script src="/_next/static/chunks/main.js?ts=1616720294380"></script><script src="/_next/static/chunks/webpack.js?ts=1616720294380"></script><script src="/_next/static/chunks/pages/_app.js?ts=1616720294380"></script><script src="/_next/static/chunks/pages/signup.js?ts=1616720294380"></script><script src="/_next/static/development/_buildManifest.js?ts=1616720294380"></script><script src="/_next/static/development/_ssgManifest.js?ts=1616720294380"></script>
   </body>
</html>
```

**Reply by balazsorban44** [reply]
You cannot get the CSRF token inside `getStaticProps`. `getStaticProps` does not have access to cookies.

You can go about this much easier by simply using the [`signIn("credentials", {username, password})`](https://next-auth.js.org/getting-started/client#signin) method.

**Reply by jayliew** [reply]
Solved the problem. I changed to using `getServerSideProps` and it worked as expected.

I've created a pull request to update code example in the documentation here: https://github.com/nextauthjs/next-auth/pull/1592

**Reply by balazsorban44** [reply]
I still suggest you rethink if you want to server-side render the whole page just to achieve login. you could try my above mentioned client-side method `signIn`. Although if you are aware of the consequences I'm fine with getServerSideProps

**Reply by jayliew** [reply]
I'm new to NextJS, so I'm trying to wrap my head around that, while now also trying to wrap my head around Next Auth, and also TypeORM 🤕 

My understanding is that since a sign-in page is going to need a CSRF token that is computed at request time, there's no way this can be statically generated at build time.

I don't understand what you're saying about server side vs. client side.

My goal isn't really a custom sign in page. This was purely me trying to accomplish a dry-run of how to programmatically log-in a user. 

My real goal is to build a signup / registration page, where a user will enter a username and password. Upon successful registration, I want to programmatically sign-in that new user.

**Reply by balazsorban44** [reply]
I understand the seemingly occurring complexity with server vs. client, but for your usecase, this is exactly what the signin function will help you with!

Please have a look at the documentation https://next-auth.js.org/getting-started/client#signin

you are totally right that CSRF tokens can't be statically generated, but the above mentioned method handles CSRF for you, so you don't have to worry about it.

My point is that when you use getServerSideProps, you actually tell next.js that whatever page you are on, should be rendered each and every time for every user. if you are building a signin/registration page it sounds like this won't change often, and so getServerSideProps may be unnecessary.

the signIn function will also handle the whole login process, and in case of a successful login the user will automatically be logged in. you can check out the `redirect:false` option for better UX at the above mentioned link, if you will be using a credentials or email provider

**Comment by StringKe** [comment]
I tried to complete the automatic login with the `signIn` API, but the request returned only the url and did not complete the login. csrf is completed automatically by `signIn`.

What should I change?


```tsx
<form
                    onSubmit={form.onSubmit((values) => {
                        onRegistry(values)
                            .then((response) => {
                                if (response.ok) {
                                    showNotification({
                                        title: 'Success',
                                        message:
                                            'Sign Up successful, automatic login after a few seconds',
                                        color: 'green',
                                        onClose: () => {
                                            void signIn('credentials', {
                                                redirect: false,
                                                email: values.email,
                                                password: values.password,
                                            })
                                                .then((response) => {
                                                    if (
                                                        response &&
                                                        response?.ok
                                                    ) {
                                                        void router.push('/')
                                                    } else {
                                                        void router.push(
                                                            '/auth/member/login',
                                                        )
                                                    }
                                                })
                                                .catch(() => {
                                                    void router.push(
                                                        '/auth/member/login',
                                                    )
                                                })
                                        },
                                    })
                                } else {
                                    showNotification({
                                        title: 'Error',
                                        message: response.message,
                                        color: 'red',
                                    })
                                }
                            })
                            .catch(() => {
                                showNotification({
                                    title: 'Error',
                                    message: 'Sign Up failed',
                                    color: 'red',
                                })
                            })
                    })}
                    className={classes.form}
                >
                ....
```
